src/phoenix/datasets/types.py

```diff
@@ -165,48 +226,142 @@ def __post_init__(self) -> None:
             ValueError("Must specify either result or error")
 
 
-class _HasName(Protocol):
-    name: str
+ExampleOutput: TypeAlias = Mapping[str, JSONSerializable]
+ExampleMetadata: TypeAlias = Mapping[str, JSONSerializable]
+ExampleInput: TypeAlias = Mapping[str, JSONSerializable]
 
+EvaluatorName: TypeAlias = str
+EvaluatorKind: TypeAlias = str
+EvaluatorOutput: TypeAlias = Union[EvaluationResult, bool, int, float, str]
+
+
+class Evaluator(ABC):
+    _kind: EvaluatorKind
+    _name: EvaluatorName
 
-class _HasKind(Protocol):
     @property
-    def annotator_kind(self) -> str: ...
+    def name(self) -> EvaluatorName:
+        if hasattr(self, "_name"):
+            return self._name
+        return self.__class__.__name__
 
+    @property
+    def kind(self) -> EvaluatorKind:
+        if hasattr(self, "_kind"):
+            return self._kind
+        return "CODE"
+
+    def __new__(cls, *args: Any, **kwargs: Any) -> Evaluator:
+        if cls is Evaluator:
+            raise TypeError(f"{cls.__name__} should not be instantiated.")
+        return object.__new__(cls)
 
-@runtime_checkable
-class CanEvaluate(_HasName, _HasKind, Protocol):
     def evaluate(
         self,
-        example: Example,
-        experiment_run: ExperimentRun,
-    ) -> EvaluationResult: ...
-
+        *,
+        output: TaskOutput,
+        expected: ExampleOutput,
+        metadata: ExampleMetadata,
+        input: ExampleInput,
+        **kwargs: Any,
+    ) -> EvaluationResult:
+        raise NotImplementedError
 
-@runtime_checkable
-class CanAsyncEvaluate(_HasName, _HasKind, Protocol):
     async def async_evaluate(
         self,
-        example: Example,
-        experiment_run: ExperimentRun,
-    ) -> EvaluationResult: ...
+        *,
+        output: TaskOutput,
+        expected: ExampleOutput,
+        metadata: ExampleMetadata,
+        input: ExampleInput,
+        **kwargs: Any,
+    ) -> EvaluationResult:
+        return self.evaluate(
+            output=output,
+            expected=expected,
+            metadata=metadata,
+            input=input,
+            **kwargs,
```

Comment:
If they don't implement either, there'll be an error message that tells them to implement either one. There's no restriction on which one they have to implement.

Reply:
I understand. I just think it would be helpful to have some docstrings to help users understand which methods to sub-class and when. Non-blocking if we are planning to add that later.